apis/Google.Cloud.Firestore.Data/docs/userguide.md

```diff
@@ -0,0 +1,127 @@
+
+Specific fields in the document can be modified using `DocumentReference.UpdateAsync`. Currently this
+requires a `Dictionary<FieldMask, object>` to be passed in, but we intend to improve this experience.
+See [further work items](furtherwork.md) for more discussion over improving this.
+
+A precondition may be specified for the update.
+
+[!code-cs[](obj/snippets/Google.Cloud.Firestore.Data.UserGuide.txt#UpdateAsync)]
+
+### Setting document data with optional merging
+
+The `SetAsync` operation is versatile. By default it replaces all data in the document. For example, this code:
+
+[!code-cs[](obj/snippets/Google.Cloud.Firestore.Data.UserGuide.txt#SetAsyncOverwrite)]
+
+... would end up wiping out setting the `State` field in the document to a null value. (`Capital` would still be `false`,
+as that's the default value of the `bool` type.)
+
+To use our class model but specify which fields we want to merge, we can specify an appropriate `SetOptions`.
+
+[!code-cs[](obj/snippets/Google.Cloud.Firestore.Data.UserGuide.txt#SetAsyncMergeSpecific)]
+
+A simpler option in many cases - particularly when your document doesn't deal with nested data - is to use an anonymous
+type to specify the fields you want to modify, and specify `SetOptions.MergeAll` so that all the fields you've specified
+in the anonymous type are merged, but no others.
+
+[!code-cs[](obj/snippets/Google.Cloud.Firestore.Data.UserGuide.txt#SetAsyncMergeAll)]
+
+### Deleting a document
+
+Deleting a document is simple via `DocumentReference.DeleteAsync`. A precondition can be specified (for example, to only
+delete the document if its last-update timestamp matches one you know); otherwise the delete operation is unconditional.
+
+[!code-cs[](obj/snippets/Google.Cloud.Firestore.Data.UserGuide.txt#DeleteAsync)]
+
+All of these operations can also be performed in batches via `WriteBatch`, or within transactions.
+
+## Reading documents
+
+Once your data is in Firestore, you probably want to read it at some point.
+
+### Fetching a document snapshot
+
+`DocumentReference` allows you to fetch a snapshot of a document:
+
+[!code-cs[](obj/snippets/Google.Cloud.Firestore.Data.UserGuide.txt#DocumentSnapshot)]
+
+### Querying
+
+You can also query collections, either directly to retrieve all the data from all the documents in the collection,
+or with filtering, projections, ordering etc.
+
+[!code-cs[](obj/snippets/Google.Cloud.Firestore.Data.UserGuide.txt#QuerySnapshot)]
+
+## Transactions
+
+Transactions accept a callback of user code, which is then passed a `Transaction` object
+to work with.
+
+Transactions always run asynchronously, but may be given a synchronous callback;
+as modification operations (`Create`, `Update` etc) are specified synchronously and batched locally,
+transactions often won't need any asynchronous code. However, the operations to fetch document or
+query snapshots within a transaction do require asynchrony.
+
+The callback can return a value, but doesn't have to.
```

Comment:
Slightly clumsy. "The callback can optionally return a value" perhaps?

Reply:
Done.